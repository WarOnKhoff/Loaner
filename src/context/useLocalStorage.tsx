import React, { useState, useEffect } from "react"

export function useLocalStorage(
  localStorageKey: string,
  defaultValue?: number
): [number, React.Dispatch<number>] {
  const [value, setValue] = useState<number>(
    Number(localStorage.getItem(localStorageKey)) || defaultValue || 0
  )
  useEffect(() => {
    localStorage.setItem(localStorageKey, value.toString())
  }, [value, localStorageKey])

  return [value, setValue]
}

export function useLocalStorageTheme(
  localStorageKey: string
): [string, React.Dispatch<string>] {
  const [value, setValue] = useState<string>(
    localStorage.getItem(localStorageKey) || "dark"
  )

  useEffect(() => {
    localStorage.setItem(localStorageKey, value.toString())
  }, [value, localStorageKey])
  return [value, setValue]
}
