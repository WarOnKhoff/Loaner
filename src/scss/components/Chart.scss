.chart {
  margin-top: 5rem;
  width: 100%;
  padding: 3rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  border: 1px solid rgba(0, 0, 0, 0.5);
  box-sizing: border-box;
  box-shadow: 4px 8px 3px rgba(0, 0, 0, 0.25);
  border-radius: 10px;
  background-color: $lightColor;
  @media (max-width: $bp-medium) {
    margin-top: 0;
  }
  &_btn {
    outline: none;
    text-decoration: none;
    border: none;
    width: 18rem;
    height: 6rem;
    background-color: $secondaryColor;
    color: $lightColor;
    border-radius: 10px;
    display: flex;
    justify-content: center;
    align-items: center;
    transition: all 0.3s;
    cursor: pointer;
    &:active:focus {
      outline: none;
      text-decoration: none;
      transform: scale(0.9);
    }
  }
}

.headerText {
  width: 100%;
  margin-bottom: 5rem;
  &_1 {
    margin: 0;
    width: 100%;
    text-align: start;
    font-size: 2.4rem;
    color: $secondaryColor;
  }
  &_2 {
    margin: 0;
    width: 100%;
    text-align: start;
    font-size: 4.8rem;
    color: $primaryColor;
  }
}
.summary {
  display: flex;
  width: 100%;
  justify-content: space-between;
  margin-bottom: 5rem;
  &_group {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    &_text {
      font-size: 2.2rem;
      font-weight: bold;
      color: $primaryColor;
    }
    &_flex {
      display: flex;
      align-items: center;
    }
    &_value {
      font-weight: bold;
      font-size: 3.6rem;
      color: $secondaryColor;
    }
    &_postfix {
      font-size: 3rem;
      font-weight: bold;
      color: $primaryColor;
      margin-left: 1rem;
    }
  }
}

.slider_wrapper {
  width: 90%;
  margin-bottom: 5rem;
}
.slider {
  appearance: none;
  box-shadow: 0px 1px 10px 0px #333;
  height: 5rem;
  width: 100%;
  height: 2rem;
  border-radius: 5px;
  outline: none;
  &::-webkit-slider-thumb {
    appearance: none;
    display: block;
    width: 3rem;
    height: 3rem;
    border-radius: 30%;
    background-color: $primaryColor;
  }
  &:active:focus {
    outline: none;
    border: none;
  }
}

.loan_input {
  position: relative;
  &::before {
    content: "100 EUR";
    font-size: 2.5rem;
    color: #333;
    position: absolute;
    display: block;
    top: 130%;
    left: 0%;
  }
  &::after {
    content: "10000 EUR";
    font-size: 2.5rem;
    color: #333;
    position: absolute;
    display: block;
    top: 130%;
    right: 0%;
  }
}
.therm_input {
  position: relative;
  &::before {
    content: "1 month";
    font-size: 2.5rem;
    color: #333;
    position: absolute;
    display: block;
    top: 130%;
    left: 0%;
  }
  &::after {
    content: "48 months";
    font-size: 2.5rem;
    color: #333;
    position: absolute;
    display: block;
    top: 130%;
    right: 0%;
  }
}
