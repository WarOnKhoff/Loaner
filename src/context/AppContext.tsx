import React, { useState } from "react"
import { useLocalStorage } from "./useLocalStorage"

interface ContextProps {
  loanAmount: number
  setLoanAmount(value: number): void
  loanTerm: number
  setLoanTerm(value: number): void
  page: string
  setPage(value: string): void
}

export const AppContext = React.createContext<ContextProps>({
  loanAmount: 100,
  setLoanAmount: () => {},
  loanTerm: 1,
  setLoanTerm: () => {},
  page: "Home",
  setPage: () => {}
})

export const AppContextProvider = (props: React.PropsWithChildren<{}>) => {
  const [loanAmount, setLoanAmount] = useLocalStorage("loan", 100)
  const [loanTerm, setLoanTerm] = useLocalStorage("term", 1)
  const [page, setPage] = useState<string>(
    window.location.pathname.substring(1)
  )
  const value = {
    loanAmount,
    setLoanAmount,
    loanTerm,
    setLoanTerm,
    page,
    setPage
  }
  return (
    <AppContext.Provider value={value}>{props.children}</AppContext.Provider>
  )
}
