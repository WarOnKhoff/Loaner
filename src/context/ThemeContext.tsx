import React from "react"
import { useLocalStorageTheme } from "./useLocalStorage"

interface ContextProps {
  theme: string
  themeToggler(): void
}
export const ThemeContext = React.createContext<ContextProps>({
  theme: "dark",
  themeToggler: () => {}
})

export const ThemeContextProvider = (props: React.PropsWithChildren<{}>) => {
  const [theme, setTheme] = useLocalStorageTheme("theme")
  const themeToggler = () => {
    theme === "dark" ? setTheme("light") : setTheme("dark")
  }
  const value = {
    theme,
    themeToggler
  }

  return (
    <ThemeContext.Provider value={value}>
      {props.children}
    </ThemeContext.Provider>
  )
}
