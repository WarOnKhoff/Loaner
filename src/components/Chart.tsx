import React, { useContext, useCallback, ReactElement } from "react"
import { AppContext } from "../context/AppContext"
import Slider from "./Slider"
import ValueCaption from "./ValueCaption"

const Chart = (): ReactElement => {
  const { loanAmount, setLoanAmount, loanTerm, setLoanTerm } = useContext(
    AppContext
  )
  const handleAmountChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setLoanAmount(Number(e.target.value))

  const handleTermChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setLoanTerm(Number(e.target.value))

  const getMonthlyPayment = useCallback(() => {
    // dumb monthly payment calculations
    return Math.round(loanAmount / loanTerm + (loanAmount / loanTerm) * 0.3)
  }, [loanTerm, loanAmount])

  return (
    <div className="chart" data-testid="chart">
      <div className="headerText">
        <h3 className="headerText_1">How much money do you need?</h3>
        <h2 className="headerText_2">Calculate your loan</h2>
      </div>
      <div className="slider_wrapper">
        <Slider
          minValue={100}
          maxValue={10000}
          currentValue={loanAmount}
          onChange={handleAmountChange}
          selector="loan_input"
        />
      </div>
      <div className="slider_wrapper">
        <Slider
          minValue={1}
          maxValue={48}
          currentValue={loanTerm}
          onChange={handleTermChange}
          selector="therm_input"
        />
      </div>
      <div className="summary">
        <ValueCaption
          caption="You are getting"
          value={loanAmount}
          postfix="EUR"
          testId="loan_value"
        />
        <ValueCaption
          caption="Terms of use"
          value={loanTerm}
          postfix={loanTerm > 1 ? "months" : "month"}
          testId="therm_value"
        />
        <ValueCaption
          caption="Monthly payment"
          value={getMonthlyPayment()}
          postfix="EUR"
        />
      </div>
      <button className="chart_btn">Apply</button>
    </div>
  )
}

export default Chart
