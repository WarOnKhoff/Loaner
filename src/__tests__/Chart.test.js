import React from "react"
import { render, fireEvent } from "@testing-library/react"
import Chart from "../components/Chart"
import "@testing-library/jest-dom/extend-expect"
import { AppContextProvider } from "../context/AppContext"
const setup = () => {
  const { getByTestId, container } = render(<Chart />)
  return {
    getByTestId,
    container
  }
}

describe("Components: Chart", () => {
  test("Renders correctly", () => {
    const { container } = render(<Chart />)
    expect(container).toMatchSnapshot()
  })
  test("Should keep 5 children nodes", () => {
    expect(setup().getByTestId("chart").childElementCount).toBe(5)
  })
  test("Should change value of ValueCaption on Slider component onChange event", () => {
    const { container, getByTestId } = render(
      <AppContextProvider>
        <Chart />
      </AppContextProvider>
    )
    const loanInput = getByTestId("loan_input")
    const loanValue = getByTestId("loan_value")

    const thermInput = getByTestId("therm_input")
    const thermValue = getByTestId("therm_value")

    // Loan amount input/value couple
    fireEvent.change(loanInput, { target: { value: 5 } })
    expect(loanValue.innerHTML).toBe("5")
    // Loan therm input/value couple
    fireEvent.change(thermInput, { target: { value: 10 } })
    expect(thermValue.innerHTML).toBe("10")
  })
})
