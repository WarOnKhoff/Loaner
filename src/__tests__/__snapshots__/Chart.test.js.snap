// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Chart Renders correctly 1`] = `
<div>
  <div
    class="chart"
    data-testid="chart"
  >
    <div
      class="headerText"
    >
      <h3
        class="headerText_1"
      >
        How much money do you need?
      </h3>
      <h2
        class="headerText_2"
      >
        Calculate your loan
      </h2>
    </div>
    <div
      class="slider_wrapper"
    >
      <input
        class="slider loan_input"
        data-testid="loan_input"
        max="10000"
        min="100"
        type="range"
        value="100"
      />
    </div>
    <div
      class="slider_wrapper"
    >
      <input
        class="slider therm_input"
        data-testid="therm_input"
        max="48"
        min="1"
        type="range"
        value="1"
      />
    </div>
    <div
      class="summary"
    >
      <div
        class="summary_group"
      >
        <div
          class="summary_group_text"
          data-testid="capture_text"
        >
          You are getting
        </div>
        <div
          class="summary_group_flex"
        >
          <div
            class="summary_group_value"
            data-testid="loan_value"
          >
            100
          </div>
          <div
            class="summary_group_postfix"
            data-testid="capture_postfix"
          >
            EUR
          </div>
        </div>
      </div>
      <div
        class="summary_group"
      >
        <div
          class="summary_group_text"
          data-testid="capture_text"
        >
          Terms of use
        </div>
        <div
          class="summary_group_flex"
        >
          <div
            class="summary_group_value"
            data-testid="therm_value"
          >
            1
          </div>
          <div
            class="summary_group_postfix"
            data-testid="capture_postfix"
          >
            month
          </div>
        </div>
      </div>
      <div
        class="summary_group"
      >
        <div
          class="summary_group_text"
          data-testid="capture_text"
        >
          Monthly payment
        </div>
        <div
          class="summary_group_flex"
        >
          <div
            class="summary_group_value"
          >
            130
          </div>
          <div
            class="summary_group_postfix"
            data-testid="capture_postfix"
          >
            EUR
          </div>
        </div>
      </div>
    </div>
    <button
      class="chart_btn"
    >
      Apply
    </button>
  </div>
</div>
`;
